### DISCUSSION TITLE
The Responder is accepting the connection from the initiator even after the connection is terminated and unloaded in the responder

### DISCUSSION THREAD

**Discussion by Kannan-1989** [strongswan/strongswan#2294] [Q&A] [answered]
Initially, few IPsec sessions are established between the initiator and the responder device. 
Below are the configurations (not everything mentioned here) used,

**Initiator Device:**
DpdAction = "restart"
StartAction = "start"
CloseAction = "start"

**Responder Device:**
DpdAction = "clear"
CloseAction = "none"
StartAction = "none"

**In Both Devices:**
DpdDelay:    "40s"
KeyingTries: "0"
Encap:       "yes"
Mobike:      "no"

After some time, we are calling the VICI Terminate and UnloadConnection APIs in the responder for all the connections. In the terminate API, we are setting only force= TRUE and ike = Corresponding IKE_SA Name.
Some of the sessions are getting terminated in the responder as expected but some connections are again re-established in the responder. 

**We can understand that the below is happening:**
In the responder, when we call Terminate and Unload APIs, it will send a delete message to the initiator. On receiving the delete message, due to closeAction set to "start" in strongswan, it sends the Init message and some tunnels in the responder accept it and establish the session. 

**Expectation:**
Since the connections are terminated and unloaded in the responder, it should not accept and re-establish such connections from the initiator.
The responder device should establish the connections only if they are loaded in it. 

**Charon Logs from the responder:**
_tunnel ipsec-tunnel-18 (working as expected):_

May 22 09:04:11 08[CFG] vici terminate IKE_SA 'tunnel ipsec-tunnel-18'
May 22 09:04:11 09[MGR] checkout IKE_SA by unique ID 2687
May 22 09:04:11 09[MGR] IKE_SA tunnel ipsec-tunnel-18[2687] successfully checked out
May 22 09:04:11 09[IKE] <tunnel ipsec-tunnel-18|2687> queueing IKE_DELETE task
May 22 09:04:11 09[IKE] <tunnel ipsec-tunnel-18|2687> activating new tasks
May 22 09:04:11 09[IKE] <tunnel ipsec-tunnel-18|2687>   activating IKE_DELETE task
May 22 09:04:11 09[IKE] <tunnel ipsec-tunnel-18|2687> deleting IKE_SA tunnel ipsec-tunnel-18[2687] between 2001:101:202:1::1[dc2_loop18]...1:1:1:1::18[dc1_loop18]
May 22 09:04:11 09[IKE] <tunnel ipsec-tunnel-18|2687> IKE_SA tunnel ipsec-tunnel-18[2687] state change: ESTABLISHED => DELETING
May 22 09:04:11 09[IKE] <tunnel ipsec-tunnel-18|2687> sending DELETE for IKE_SA tunnel ipsec-tunnel-18[2687]
May 22 09:04:11 09[ENC] <tunnel ipsec-tunnel-18|2687> generating INFORMATIONAL request 74 [ D ]
May 22 09:04:11 09[NET] <tunnel ipsec-tunnel-18|2687> sending packet: from 2001:101:202:1::1[4500] to 1:1:1:1::18[4500] (65 bytes)
May 22 09:04:11 06[NET] sending packet: from 2001:101:202:1::1[4500] to 1:1:1:1::18[4500]
May 22 09:04:11 09[MGR] <tunnel ipsec-tunnel-18|2687> checkin and destroy IKE_SA tunnel ipsec-tunnel-18[2687]
May 22 09:04:11 09[IKE] <tunnel ipsec-tunnel-18|2687> IKE_SA tunnel ipsec-tunnel-18[2687] state change: DELETING => DESTROYING
May 22 09:04:11 09[CHD] <tunnel ipsec-tunnel-18|2687> CHILD_SA tunnel ipsec-tunnel-18{42} state change: INSTALLED => DESTROYING
May 22 09:04:11 09[KNL] <tunnel ipsec-tunnel-18|2687> deleting SAD entry with SPI ce441d60
May 22 09:04:11 09[KNL] <tunnel ipsec-tunnel-18|2687> deleted SAD entry with SPI ce441d60
May 22 09:04:11 09[KNL] <tunnel ipsec-tunnel-18|2687> deleting SAD entry with SPI c05ac16e
May 22 09:04:11 02[NET] received packet: from 1:1:1:1::18[500] to 2001:101:202:1::1[500]
May 22 09:04:11 02[NET] waiting for data on sockets
May 22 09:04:11 02[NET] received packet: from 1:1:1:1::18[4500] to 2001:101:202:1::1[4500]
May 22 09:04:11 02[NET] waiting for data on sockets
May 22 09:04:11 13[MGR] checkout IKEv2 SA by message with SPIs ff50991cd4b7beac_i 08a94ffc0a08b203_r
May 22 09:04:11 13[MGR] IKE_SA checkout not successful
May 22 09:04:11 13[MGR] checkout IKEv2 SA by message with SPIs 48eea517a638261d_i 0000000000000000_r
May 22 09:04:11 13[MGR] created IKE_SA (unnamed)[6683]
May 22 09:04:11 13[NET] <6683> received packet: from 1:1:1:1::18[500] to 2001:101:202:1::1[500] (472 bytes)
May 22 09:04:11 13[ENC] <6683> parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
May 22 09:04:11 13[IKE] <6683> local endpoint changed from 0.0.0.0[500] to 2001:101:202:1::1[500]
May 22 09:04:11 13[IKE] <6683> remote endpoint changed from 0.0.0.0 to 1:1:1:1::18[500]
May 22 09:04:11 13[IKE] <6683> 1:1:1:1::18 is initiating an IKE_SA
May 22 09:04:11 13[IKE] <6683> IKE_SA (unnamed)[6683] state change: CREATED => CONNECTING
May 22 09:04:11 13[CFG] <6683> selected proposal: IKE:AES_GCM_16_256/PRF_HMAC_SHA2_256/MODP_2048_256
May 22 09:04:11 09[KNL] <tunnel ipsec-tunnel-18|2687> deleted SAD entry with SPI c05ac16e
May 22 09:04:11 09[MGR] checkin and destroy of IKE_SA successful
May 22 09:04:11 13[IKE] <6683> remote host is behind NAT
May 22 09:04:11 13[ENC] <6683> generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
May 22 09:04:11 13[NET] <6683> sending packet: from 2001:101:202:1::1[500] to 1:1:1:1::18[500] (464 bytes)
May 22 09:04:11 06[NET] sending packet: from 2001:101:202:1::1[500] to 1:1:1:1::18[500]
May 22 09:04:11 13[MGR] <6683> checkin IKEv2 SA (unnamed)[6683] with SPIs 48eea517a638261d_i 782d56709aec6ab7_r
May 22 09:04:11 13[MGR] <6683> checkin of IKE_SA successful
May 22 09:04:11 02[NET] received packet: from 1:1:1:1::18[4500] to 2001:101:202:1::1[4500]
May 22 09:04:11 02[NET] waiting for data on sockets
May 22 09:04:11 09[MGR] checkout IKEv2 SA by message with SPIs 48eea517a638261d_i 782d56709aec6ab7_r
May 22 09:04:11 09[MGR] IKE_SA (unnamed)[6683] successfully checked out
May 22 09:04:11 09[NET] <6683> received packet: from 1:1:1:1::18[4500] to 2001:101:202:1::1[4500] (249 bytes)
May 22 09:04:11 09[ENC] <6683> parsed IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
May 22 09:04:11 09[IKE] <6683> local endpoint changed from 2001:101:202:1::1[500] to 2001:101:202:1::1[4500]
May 22 09:04:11 09[IKE] <6683> remote endpoint changed from 1:1:1:1::18[500] to 1:1:1:1::18[4500]
May 22 09:04:11 09[CFG] <6683> looking for peer configs matching 2001:101:202:1::1[dc2_loop18]...1:1:1:1::18[dc1_loop18]
May 22 09:04:11 09[CFG] <6683> **no matching peer config found**
May 22 09:04:11 09[ENC] <6683> generating IKE_AUTH response 1 [ N(AUTH_FAILED) ]
May 22 09:04:11 09[NET] <6683> sending packet: from 2001:101:202:1::1[4500] to 1:1:1:1::18[4500] (65 bytes)
May 22 09:04:11 06[NET] sending packet: from 2001:101:202:1::1[4500] to 1:1:1:1::18[4500]
May 22 09:04:11 09[MGR] <6683> checkin and destroy IKE_SA (unnamed)[6683]
May 22 09:04:11 09[IKE] <6683> IKE_SA (unnamed)[6683] state change: CONNECTING => DESTROYING
May 22 09:04:11 09[MGR] checkin and destroy of IKE_SA successful

_tunnel ipsec-tunnel-19 (not working):_

May 22 09:04:11 10[CFG] vici terminate IKE_SA 'tunnel ipsec-tunnel-19'
May 22 09:04:11 05[MGR] checkout IKE_SA by unique ID 2691
May 22 09:04:11 05[MGR] IKE_SA tunnel ipsec-tunnel-19[2691] successfully checked out
May 22 09:04:11 05[IKE] <tunnel ipsec-tunnel-19|2691> queueing IKE_DELETE task
May 22 09:04:11 05[IKE] <tunnel ipsec-tunnel-19|2691> activating new tasks
May 22 09:04:11 05[IKE] <tunnel ipsec-tunnel-19|2691>   activating IKE_DELETE task
May 22 09:04:11 05[IKE] <tunnel ipsec-tunnel-19|2691> deleting IKE_SA tunnel ipsec-tunnel-19[2691] between 2001:101:202:1::1[dc2_loop19]...1:1:1:1::19[dc1_loop19]
May 22 09:04:11 05[IKE] <tunnel ipsec-tunnel-19|2691> IKE_SA tunnel ipsec-tunnel-19[2691] state change: ESTABLISHED => DELETING
May 22 09:04:11 05[IKE] <tunnel ipsec-tunnel-19|2691> sending DELETE for IKE_SA tunnel ipsec-tunnel-19[2691]
May 22 09:04:11 05[ENC] <tunnel ipsec-tunnel-19|2691> generating INFORMATIONAL request 11 [ D ]
May 22 09:04:11 05[NET] <tunnel ipsec-tunnel-19|2691> sending packet: from 2001:101:202:1::1[4500] to 1:1:1:1::19[4500] (65 bytes)
May 22 09:04:11 05[MGR] <tunnel ipsec-tunnel-19|2691> checkin and destroy IKE_SA tunnel ipsec-tunnel-19[2691]
May 22 09:04:11 05[IKE] <tunnel ipsec-tunnel-19|2691> IKE_SA tunnel ipsec-tunnel-19[2691] state change: DELETING => DESTROYING
May 22 09:04:11 05[CHD] <tunnel ipsec-tunnel-19|2691> CHILD_SA tunnel ipsec-tunnel-19{45} state change: INSTALLED => DESTROYING
May 22 09:04:11 05[KNL] <tunnel ipsec-tunnel-19|2691> deleting SAD entry with SPI c0199e12
May 22 09:04:11 06[NET] sending packet: from 2001:101:202:1::1[4500] to 1:1:1:1::19[4500]
May 22 09:04:11 02[NET] received packet: from 1:1:1:1::19[500] to 2001:101:202:1::1[500]
May 22 09:04:11 02[NET] waiting for data on sockets
May 22 09:04:11 02[NET] received packet: from 1:1:1:1::19[4500] to 2001:101:202:1::1[4500]
May 22 09:04:11 02[NET] waiting for data on sockets
May 22 09:04:11 03[MGR] checkout IKEv2 SA by message with SPIs f1e2f384588254d3_i a4dfa8143c27104a_r
May 22 09:04:11 03[MGR] IKE_SA checkout not successful
May 22 09:04:11 03[MGR] checkout IKEv2 SA by message with SPIs dc583f7a3acfc7c3_i 0000000000000000_r
May 22 09:04:11 03[MGR] created IKE_SA (unnamed)[6684]
May 22 09:04:11 03[NET] <6684> received packet: from 1:1:1:1::19[500] to 2001:101:202:1::1[500] (472 bytes)
May 22 09:04:11 03[ENC] <6684> parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
May 22 09:04:11 03[IKE] <6684> local endpoint changed from 0.0.0.0[500] to 2001:101:202:1::1[500]
May 22 09:04:11 03[IKE] <6684> remote endpoint changed from 0.0.0.0 to 1:1:1:1::19[500]
May 22 09:04:11 03[IKE] <6684> 1:1:1:1::19 is initiating an IKE_SA
May 22 09:04:11 03[IKE] <6684> IKE_SA (unnamed)[6684] state change: CREATED => CONNECTING
May 22 09:04:11 03[CFG] <6684> selected proposal: IKE:AES_GCM_16_256/PRF_HMAC_SHA2_256/MODP_2048_256
May 22 09:04:11 05[KNL] <tunnel ipsec-tunnel-19|2691> deleted SAD entry with SPI c0199e12
May 22 09:04:11 05[KNL] <tunnel ipsec-tunnel-19|2691> deleting SAD entry with SPI cb38a9c5
May 22 09:04:11 05[KNL] <tunnel ipsec-tunnel-19|2691> deleted SAD entry with SPI cb38a9c5
May 22 09:04:11 05[MGR] checkin and destroy of IKE_SA successful
May 22 09:04:11 03[IKE] <6684> remote host is behind NAT
May 22 09:04:11 03[ENC] <6684> generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
May 22 09:04:11 03[NET] <6684> sending packet: from 2001:101:202:1::1[500] to 1:1:1:1::19[500] (464 bytes)
May 22 09:04:11 06[NET] sending packet: from 2001:101:202:1::1[500] to 1:1:1:1::19[500]
May 22 09:04:11 03[MGR] <6684> checkin IKEv2 SA (unnamed)[6684] with SPIs dc583f7a3acfc7c3_i 1bf1863bf44eaeae_r
May 22 09:04:11 03[MGR] <6684> checkin of IKE_SA successful
May 22 09:04:11 02[NET] received packet: from 1:1:1:1::19[4500] to 2001:101:202:1::1[4500]
May 22 09:04:11 02[NET] waiting for data on sockets
May 22 09:04:11 03[MGR] checkout IKEv2 SA by message with SPIs dc583f7a3acfc7c3_i 1bf1863bf44eaeae_r
May 22 09:04:11 03[MGR] IKE_SA (unnamed)[6684] successfully checked out
May 22 09:04:11 03[NET] <6684> received packet: from 1:1:1:1::19[4500] to 2001:101:202:1::1[4500] (249 bytes)
May 22 09:04:11 03[ENC] <6684> parsed IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
May 22 09:04:11 03[IKE] <6684> local endpoint changed from 2001:101:202:1::1[500] to 2001:101:202:1::1[4500]
May 22 09:04:11 03[IKE] <6684> remote endpoint changed from 1:1:1:1::19[500] to 1:1:1:1::19[4500]
May 22 09:04:11 03[CFG] <6684> looking for peer configs matching 2001:101:202:1::1[dc2_loop19]...1:1:1:1::19[dc1_loop19]
May 22 09:04:11 03[CFG] <tunnel ipsec-tunnel-19|6684> **selected peer config 'tunnel ipsec-tunnel-19'**
May 22 09:04:11 03[IKE] <tunnel ipsec-tunnel-19|6684> authentication of 'dc1_loop19' with pre-shared key successful
May 22 09:04:11 03[IKE] <tunnel ipsec-tunnel-19|6684> authentication of 'dc2_loop19' (myself) with pre-shared key
May 22 09:04:11 03[IKE] <tunnel ipsec-tunnel-19|6684> successfully created shared key MAC
May 22 09:04:11 03[IKE] <tunnel ipsec-tunnel-19|6684> IKE_SA tunnel ipsec-tunnel-19[6684] established between 2001:101:202:1::1[dc2_loop19]...1:1:1:1::19[dc1_loop19]
May 22 09:04:11 03[IKE] <tunnel ipsec-tunnel-19|6684> IKE_SA tunnel ipsec-tunnel-19[6684] state change: CONNECTING => ESTABLISHED
May 22 09:04:11 03[CFG] <tunnel ipsec-tunnel-19|6684> selected proposal: ESP:AES_GCM_16_256/NO_EXT_SEQ
May 22 09:04:11 03[KNL] <tunnel ipsec-tunnel-19|6684> got SPI c29c8e0c
May 22 09:04:11 03[CHD] <tunnel ipsec-tunnel-19|6684> CHILD_SA tunnel ipsec-tunnel-19{51} state change: CREATED => INSTALLING
May 22 09:04:11 03[CHD] <tunnel ipsec-tunnel-19|6684>   using AES_GCM_16 for encryption
May 22 09:04:11 03[CHD] <tunnel ipsec-tunnel-19|6684> adding inbound ESP SA
May 22 09:04:11 03[CHD] <tunnel ipsec-tunnel-19|6684>   SPI 0xc29c8e0c, src 1:1:1:1::19 dst 2001:101:202:1::1
May 22 09:04:11 03[KNL] <tunnel ipsec-tunnel-19|6684> adding SAD entry with SPI c29c8e0c and reqid {1}
May 22 09:04:11 03[KNL] <tunnel ipsec-tunnel-19|6684>   using encryption algorithm AES_GCM_16 with key size 288
May 22 09:04:11 03[KNL] <tunnel ipsec-tunnel-19|6684>   using replay window of 32 packets
May 22 09:04:11 03[KNL] <tunnel ipsec-tunnel-19|6684>   HW offload: no
May 22 09:04:11 03[CHD] <tunnel ipsec-tunnel-19|6684> adding outbound ESP SA
May 22 09:04:11 03[CHD] <tunnel ipsec-tunnel-19|6684>   SPI 0xcd22c33d, src 2001:101:202:1::1 dst 1:1:1:1::19
May 22 09:04:11 03[KNL] <tunnel ipsec-tunnel-19|6684> adding SAD entry with SPI cd22c33d and reqid {1}
May 22 09:04:11 03[KNL] <tunnel ipsec-tunnel-19|6684>   using encryption algorithm AES_GCM_16 with key size 288
May 22 09:04:11 03[KNL] <tunnel ipsec-tunnel-19|6684>   using replay window of 0 packets
May 22 09:04:11 03[KNL] <tunnel ipsec-tunnel-19|6684>   HW offload: no
May 22 09:04:11 03[IKE] <tunnel ipsec-tunnel-19|6684> CHILD_SA tunnel ipsec-tunnel-19{51} established with SPIs c29c8e0c_i cd22c33d_o and TS 0.0.0.0/0 === 0.0.0.0/0
May 22 09:04:11 03[CHD] <tunnel ipsec-tunnel-19|6684> CHILD_SA tunnel ipsec-tunnel-19{51} state change: INSTALLING => INSTALLED
May 22 09:04:11 03[ENC] <tunnel ipsec-tunnel-19|6684> generating IKE_AUTH response 1 [ IDr AUTH SA TSi TSr ]
May 22 09:04:11 03[NET] <tunnel ipsec-tunnel-19|6684> sending packet: from 2001:101:202:1::1[4500] to 1:1:1:1::19[4500] (199 bytes)
May 22 09:04:11 03[MGR] <tunnel ipsec-tunnel-19|6684> checkin IKEv2 SA tunnel ipsec-tunnel-19[6684] with SPIs dc583f7a3acfc7c3_i 1bf1863bf44eaeae_r
May 22 09:04:11 03[MGR] <tunnel ipsec-tunnel-19|6684> checkin of IKE_SA successful
May 22 09:04:11 06[NET] sending packet: from 2001:101:202:1::1[4500] to 1:1:1:1::19[4500]
May 22 09:04:11 09[CFG] vici terminate IKE_SA 'tunnel ipsec-tunnel-20'
May 22 09:04:11 14[MGR] checkout IKE_SA by unique ID 2693
May 22 09:04:11 14[MGR] IKE_SA tunnel ipsec-tunnel-20[2693] successfully checked out

**Comment by tobiasbrunner** [comment, marked as answer]
> After some time, we are calling the VICI Terminate and UnloadConnection APIs in the responder for all the connections.

That order is clearly not ideal if you expect the peer to reestablish the SA due to the delete. As you saw, there is a chance that the IKE_SA_INIT reaches the responder before the connection is unloaded. So you might want to unload the connection before terminating the existing SAs.

**Reply by Kannan-1989** [reply]
Ok, let me change the order of calling APIs (i.e., unload first and then terminate the tunnel) and update soon.

**Reply by Kannan-1989** [reply]
Thanks, it worked.
